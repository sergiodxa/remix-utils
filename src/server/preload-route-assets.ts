import { EntryContext } from "@remix-run/server-runtime";

type Link = { href: string; as: string };

/**
 * Combine `preloadLinkedAssets` and `preloadModuleAssets` into a single
 * function and preload both linked and module assets.
 * @param context Remix's EntryContext
 * @param headers The headers object to append the preload links to
 * @example
 * export default function handleRequest(
 *   request: Request,
 *   statusCode: number,
 *   headers: Headers,
 *   context: EntryContext
 * ) {
 *   let markup = renderToString(
 *     <RemixServer context={context} url={request.url} />
 *   );
 *   headers.set("Content-Type", "text/html");
 *
 *   preloadRouteAssets(context, headers); // add this line
 *
 *   return new Response("<!DOCTYPE html>" + markup, {
 *     status: statusCode,
 *     headers: headers,
 *   });
 * }
 */
export function preloadRouteAssets(context: EntryContext, headers: Headers) {
  preloadLinkedAssets(context, headers); // preload links
  preloadModuleAssets(context, headers); // preload JS modules
}

/**
 * Preload the assets linked in the routes matching the current request.
 * This function will preload any `<link rel="preload" />` tag added with
 * `LinksFunction` and any CSS files linked with `<link rel="stylesheet" />`.
 * @param context Remix's EntryContext
 * @param headers The headers object to append the preload links to
 * @example
 * export default function handleRequest(
 *   request: Request,
 *   statusCode: number,
 *   headers: Headers,
 *   context: EntryContext
 * ) {
 *   let markup = renderToString(
 *     <RemixServer context={context} url={request.url} />
 *   );
 *   headers.set("Content-Type", "text/html");
 *
 *   preloadLinkedAssets(context, headers); // add this line
 *
 *   return new Response("<!DOCTYPE html>" + markup, {
 *     status: statusCode,
 *     headers: headers,
 *   });
 * }
 */
export function preloadLinkedAssets(context: EntryContext, headers: Headers) {
  let links = context.staticHandlerContext.matches
    .flatMap((match) => {
      let route = context.routeModules[match.route.id];
      if (route.links instanceof Function) return route.links();
      return [];
    })
    .map((link) => {
      if ("as" in link && "href" in link) {
        return { href: link.href, as: link.as } as Link;
      }
      if ("rel" in link && "href" in link && link.rel === "stylesheet")
        return { href: link.href, as: "style" } as Link;
      return null;
    })
    .filter((link: Link | null): link is Link => {
      return link !== null && "href" in link;
    })
    .filter((item, index, list) => {
      return index === list.findIndex((link) => link.href === item.href);
    });

  for (let link of links) {
    headers.append("Link", `<${link.href}>; rel=preload; as=${link.as}`);
  }
}

/**
 * Add Link headers to preload the JS modules in the route matching the current
 * request.
 * @param context Remix's EntryContext
 * @param headers The headers object to append the preload links to
 * @example
 * export default function handleRequest(
 *   request: Request,
 *   statusCode: number,
 *   headers: Headers,
 *   context: EntryContext
 * ) {
 *   let markup = renderToString(
 *     <RemixServer context={context} url={request.url} />
 *   );
 *   headers.set("Content-Type", "text/html");
 *
 *   preloadModuleAssets(context, headers); // add this line
 *
 *   return new Response("<!DOCTYPE html>" + markup, {
 *     status: statusCode,
 *     headers: headers,
 *   });
 * }
 */
export function preloadModuleAssets(context: EntryContext, headers: Headers) {
  let urls: string[] = [
    context.manifest.url,
    context.manifest.entry.module,
    ...context.manifest.entry.imports,
  ];

  for (let match of context.staticHandlerContext.matches) {
    let route = context.manifest.routes[match.route.id];
    urls.push(route.module, ...(route.imports ?? []));
  }

  for (let url of urls) {
    headers.append(
      "Link",
      `<${url}>; rel=preload; as=script; crossorigin=anonymous`
    );
  }
}
