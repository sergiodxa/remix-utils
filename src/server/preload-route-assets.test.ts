import { describe, expect, test } from "bun:test";
import type { EntryContext } from "react-router";
import { preloadRouteAssets } from "./preload-route-assets.js";

// TODO Fix this
describe(preloadRouteAssets, () => {
	let context: EntryContext = {
		isSpaMode: false,
		manifest: {
			url: "/manifest.js",
			routes: {
				root: {
					hasAction: false,
					hasClientAction: false,
					hasClientLoader: false,
					hasErrorBoundary: false,
					hasLoader: false,
					id: "root",
					module: "/root.js",
					imports: ["/chunk-3.js", "/chunk-4.js"],
					hasClientMiddleware: false,
					clientActionModule: undefined,
					clientLoaderModule: undefined,
					clientMiddlewareModule: undefined,
					hydrateFallbackModule: undefined,
				},
				"routes/index": {
					hasAction: false,
					hasClientAction: false,
					hasClientLoader: false,
					hasErrorBoundary: false,
					hasLoader: false,
					id: "routes/index",
					imports: ["/chunk-5.js"],
					module: "/routes/index.js",
					hasClientMiddleware: false,
					clientActionModule: undefined,
					clientLoaderModule: undefined,
					clientMiddlewareModule: undefined,
					hydrateFallbackModule: undefined,
				},
			},
			version: "latest",
			entry: {
				imports: ["/chunk-1.js", "/chunk-2.js"],
				module: "/entry.client.js",
			},
		},
		routeModules: {
			root: {
				default() {
					return null;
				},
				links() {
					return [{ rel: "stylesheet", href: "/styles.css" }];
				},
			},
			"routes/index": {
				default() {
					return null;
				},
				links: undefined,
			},
		},
		future: {
			unstable_middleware: true,
			unstable_subResourceIntegrity: true,
		},
		staticHandlerContext: {
			actionData: {},
			actionHeaders: {},
			basename: "/",
			errors: {},
			loaderData: {},
			loaderHeaders: {},
			location: {
				hash: "hash",
				key: "key",
				pathname: "/",
				search: "",
				state: null,
			},
			matches: [
				{
					params: {},
					pathname: "/",
					pathnameBase: "/",
					route: {
						id: "root",
					},
				},
				{
					params: {},
					pathname: "/",
					pathnameBase: "/",
					route: {
						id: "routes/index",
					},
				},
			],
			statusCode: 200,
		},
		serializeError(error) {
			return { message: error.message, stack: error.stack };
		},
		ssr: false,
		routeDiscovery: {
			mode: "lazy",
			manifestPath: "",
		},
	};

	test("should add the Link headers", () => {
		let headers = new Headers();

		preloadRouteAssets(context, headers);

		expect(headers.get("link")).toMatchSnapshot(
			`"</styles.css>; rel=preload; as=style, </manifest.js>; rel=preload; as=script; crossorigin=anonymous, </entry.client.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-1.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-2.js>; rel=preload; as=script; crossorigin=anonymous, </root.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-3.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-4.js>; rel=preload; as=script; crossorigin=anonymous, </routes/index.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-5.js>; rel=preload; as=script; crossorigin=anonymous"`,
		);
	});

	test("should respect manually added Link headers", () => {
		let headers = new Headers();
		headers.set("Link", "</custom.js>; rel=preload; as=script");
		preloadRouteAssets(context, headers);
		expect(headers.get("link")).toMatchSnapshot(
			`"</custom.js>; rel=preload; as=script, </styles.css>; rel=preload; as=style, </manifest.js>; rel=preload; as=script; crossorigin=anonymous, </entry.client.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-1.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-2.js>; rel=preload; as=script; crossorigin=anonymous, </root.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-3.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-4.js>; rel=preload; as=script; crossorigin=anonymous, </routes/index.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-5.js>; rel=preload; as=script; crossorigin=anonymous"`,
		);
	});
});
