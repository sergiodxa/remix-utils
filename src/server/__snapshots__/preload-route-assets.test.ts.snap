// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`preloadRouteAssets should add the Link headers: "</styles.css>; rel=preload; as=style, </manifest.js>; rel=preload; as=script; crossorigin=anonymous, </entry.client.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-1.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-2.js>; rel=preload; as=script; crossorigin=anonymous, </root.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-3.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-4.js>; rel=preload; as=script; crossorigin=anonymous, </routes/index.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-5.js>; rel=preload; as=script; crossorigin=anonymous" 1`] = `"</styles.css>; rel=preload; as=style, </manifest.js>; rel=preload; as=script; crossorigin=anonymous, </entry.client.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-1.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-2.js>; rel=preload; as=script; crossorigin=anonymous, </root.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-3.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-4.js>; rel=preload; as=script; crossorigin=anonymous, </routes/index.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-5.js>; rel=preload; as=script; crossorigin=anonymous"`;

exports[`preloadRouteAssets should respect manually added Link headers: "</custom.js>; rel=preload; as=script, </styles.css>; rel=preload; as=style, </manifest.js>; rel=preload; as=script; crossorigin=anonymous, </entry.client.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-1.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-2.js>; rel=preload; as=script; crossorigin=anonymous, </root.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-3.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-4.js>; rel=preload; as=script; crossorigin=anonymous, </routes/index.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-5.js>; rel=preload; as=script; crossorigin=anonymous" 1`] = `"</custom.js>; rel=preload; as=script, </styles.css>; rel=preload; as=style, </manifest.js>; rel=preload; as=script; crossorigin=anonymous, </entry.client.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-1.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-2.js>; rel=preload; as=script; crossorigin=anonymous, </root.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-3.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-4.js>; rel=preload; as=script; crossorigin=anonymous, </routes/index.js>; rel=preload; as=script; crossorigin=anonymous, </chunk-5.js>; rel=preload; as=script; crossorigin=anonymous"`;
